contracts/Pool.sol

```diff
@@ -349,6 +547,73 @@ abstract contract Pool {
         amtDeltas[thisCycle].thisCycle += thisCycleBlocks * perBlockDelta;
         amtDeltas[thisCycle].nextCycle += nextCycleBlocks * perBlockDelta;
     }
+
+    /// @notice Stops proxy of `msg.sender` for the duration of the modified function.
+    /// This removes and then restores any effects of the proxy on all of its receivers' futures.
+    /// It allows the function to safely modify any properties of the proxy
+    /// without having to updating the state of its receivers.
+    modifier suspendProxy {
+        applyProxyDeltasOnReceivers(-1);
+        _;
+        applyProxyDeltasOnReceivers(1);
+    }
+
+    /// @notice Applies the effects of the proxy on all of its receivers' futures.
+    /// Effects are applied as if the change was made on the beginning of the current cycle.
+    /// @param multiplier The multiplier of the deltas applied on the receivers.
+    /// `-1` to remove the effects of the proxy, `1` to reapply after removal.
+    function applyProxyDeltasOnReceivers(int8 multiplier) internal {
+        uint64 blockNumber = uint64(block.number);
+        Proxy storage proxy = proxies[msg.sender];
+        uint32 receiversCount = 0;
+        // Create an in-memory copy of the receivers list to reduce storage access
+        ReceiverWeight[] memory receiversList = new ReceiverWeight[](PROXY_WEIGHTS_COUNT_MAX);
+        // Iterating over receivers, see `ReceiverWeights` for details
+        address receiverAddr = ReceiverWeightsImpl.ADDR_ROOT;
+        while (true) {
+            uint32 weight;
+            (receiverAddr, weight, ) = proxy.receiverWeights.nextWeightPruning(receiverAddr);
+            if (receiverAddr == ReceiverWeightsImpl.ADDR_ROOT) break;
+            require(receiversCount < PROXY_WEIGHTS_COUNT_MAX, "Too many proxy receivers");
+            receiversList[receiversCount++] = ReceiverWeight(receiverAddr, weight);
+        }
+        // The proxy doesn't exist
+        if (receiversCount == 0) return;
+
+        // Iterating over deltas, see `ProxyDeltas` for details
+        uint64 cycle = ProxyDeltasImpl.CYCLE_ROOT;
+        uint64 finishedCycle = blockNumber / cycleBlocks;
+        uint64 currCycle = finishedCycle + 1;
+        int128 totalDelta = 0;
+        while (true) {
+            int128 thisCycleDelta;
+            int128 nextCycleDelta;
+            (cycle, thisCycleDelta, nextCycleDelta) = proxy.amtPerWeightDeltas.nextDeltaPruning(
+                cycle,
+                finishedCycle
+            );
+            if (cycle > finishedCycle) thisCycleDelta *= multiplier;
+            else thisCycleDelta = 0;
+            nextCycleDelta *= multiplier;
+            if (cycle == ProxyDeltasImpl.CYCLE_ROOT) break;
```

Comment:
Can we re-arrange the `if` statements to make them easier to follow? The `if/break` statement here should come as early as possible. Then it should also be possible to merge the two remaining `if` statements.

Reply:
Yeah, I took a look at this part and it was a completely uncleaned development mess, fixed :grimacing: